Assert equal measurement observable

Tanuj-rai opened the pull request:
### Before submitting

Please complete the following checklist when submitting a PR:

- [ ] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      test directory!

- [ ] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [ ] Ensure that the test suite passes, by running `make test`.

- [ ] Add a new entry to the `doc/releases/changelog-dev.md` file, summarizing the
      change, and including a link back to the PR.

- [ ] The PennyLane source code conforms to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      We check all of our code against [Pylint](https://www.pylint.org/).
      To lint modified files, simply `pip install pylint`, and then
      run `pylint pennylane/path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**
qml.assert_equal() does not provide helpful error messages when comparing unequal measurement results, making debugging more difficult during testing.


**Description of the Change:**
This PR added a basic check for unequal observables in MeasurementProcess to return a clear error message in qml.assert_equal().

**Benefits:**
 Makes test failures involving measurements easier to debug

**Related GitHub Issues:**
#7354

andrijapau commented on pennylane/ops/functions/equal.py:
```diff
@@ -805,3 +805,10 @@ def _equal_prep_sel_prep(
+
+@_equal_dispatch.register(MeasurementProcess, MeasurementProcess)
+def _equal_measurement_measurement(m1, m2):
+    if m1.obs != m2.obs:
+        return f"{m1} and {m2} are not equal because their observables differ."
+    return True
```
```suggestion
    return qml.equal(m1, m2)
```

andrijapau reviewed (changes requested):
Thanks for opening this PR! 🚀 

Could you add a `doc/releases/changelog-dev.md` entry summarizing this change?

andrijapau commented:
Just edited your description to have "Fixes" before the issue tag,
```
Related GitHub Issues:
Fixes #7354
```
This way, this PR is now linked to the original issue; and when its merged it will automatically close it. 😃

Tanuj-rai commented:
> Thanks for opening this PR! 🚀
> 
> Could you add a `doc/releases/changelog-dev.md` entry summarizing this change?

Thank you for the suggestion, @andrijapau! I’ve already committed the change, but I wanted to understand a bit better does using qml.equal(m1, m2) already handles cases where the observables are different and provides a helpful error message.

I have also updated the changelog-dev.md summarizing the change. Could you please take a look to see if anything needs to be improved?

andrijapau commented:
Hi @Tanuj-rai, sorry I misread the original PR issue. We currently have an [existing dispatch method](https://github.com/PennyLaneAI/pennylane/blob/1bfe8e4ed8e2aefc37e198acf7c3fdfa2533f9e4/pennylane/ops/functions/equal.py#L654) for two measurement processes. What we would like to do is modify it (see my screenshot for example) so that we get informative errors.

> In this task, you should update the [_equal_dispatch](https://github.com/PennyLaneAI/pennylane/blob/a2ee4b35665588be03b6dc9013c7867c1997175d/pennylane/ops/functions/equal.py#L654) functions for measurement processes located in [pennylane/ops/functions/equal.py](https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/ops/functions/equal.py#L654). Instead of returning True and False, it should return True or a string containing useful information.

![image](https://github.com/user-attachments/assets/4afe5e83-2236-46c9-bb3d-7709e2b7c416)

With the changes made above, you would get the desired behaviour,
```python
>>> qml.assert_equal(qml.expval(qml.X(1)), qml.expval(qml.X(0)))
AssertionError: op1 and op2 have different wires. Got Wires([1]) and Wires([0]).
```
as opposed to the current ambiguous message,
```python
AssertionError: expval(X(1)) and expval(X(0)) are not equal for an unspecified reason.
```

Let me know if that's clear. You will need to revert my suggestion change in https://github.com/PennyLaneAI/pennylane/pull/7421/commits/2eaf8dc86ac2a9e9550df77c806667d1ff76cd62 and fix the remaining functions accordingly. 😄 

Feel free to re-tag me to take a look. Please reference https://github.com/PennyLaneAI/pennylane/issues/5644 and https://github.com/PennyLaneAI/pennylane/issues/5648 as they are closed issues with example PRs that are relevant to this one.

Tanuj-rai commented:
Hey @andrijapau, thank you for reviewing and suggesting improvements. I've reverted the change and updated the _equal_dispatch function for MeasurementProcess. Please let me know if anything further needs to be updated.

andrijapau commented:
> Hey @andrijapau, thank you for reviewing and suggesting improvements. I've reverted the change and updated the _equal_dispatch function for MeasurementProcess. Please let me know if anything further needs to be updated.

Yes, the other branches in this function will need changes as well. For example this,
```python
qml.assert_equal(qml.measure(0), qml.measure(1))
```
should give an informative error message as well indicating that the measurement values are not equal (does not on master).

Tanuj-rai commented:
I'm sorry, @andrijapau, for the late response. I have improved the other branches of the _equal_dispatch function. Please take a look at them.

andrijapau reviewed (changes requested):
Looking better! 🚀 

Could you also write the relevant tests in `test_equal.py`? You can take a look at our development [guide](https://docs.pennylane.ai/en/stable/development/guide/tests.html) for best practices.

Once the tests are ready I can run our CI on them and make sure everything is 🟢 .

Tanuj-rai commented:
> Looking better! 🚀
> 
> Could you also write the relevant tests in `test_equal.py`? You can take a look at our development [guide](https://docs.pennylane.ai/en/stable/development/guide/tests.html) for best practices.
> 
> Once the tests are ready I can run our CI on them and make sure everything is 🟢 .

Thank you @andrijapau for the suggested changes. I will add tests as soon as possible.

andrijapau commented:
> > Looking better! 🚀
> > Could you also write the relevant tests in `test_equal.py`? You can take a look at our development [guide](https://docs.pennylane.ai/en/stable/development/guide/tests.html) for best practices.
> > Once the tests are ready I can run our CI on them and make sure everything is 🟢 .
> 
> Thank you @andrijapau for the suggested changes. I will add tests as soon as possible.

Amazing, thank you! If you need anything don't hesitate to reach out. 👍🏼

Tanuj-rai commented:
Hi @andrijapau, sorry for the delay. I have added some tests. Could you please review them when you have a moment?

albi3ro commented on pennylane/ops/functions/equal.py:
```diff
@@ -658,41 +658,54 @@ def _equal_measurements(
             atol=atol,
         )
+        if isinstance(dispatch_result, str):
+            return dispatch_result
```
This would indicate that the observables are not equal, but we would also want to add some context this message to indicate that this message is about measurements.

How about something like:
```suggestion
            return f"op1 and op2 have different observables because {dispatch_result}"
```

albi3ro commented on pennylane/ops/functions/equal.py:
```diff
@@ -658,41 +658,54 @@ def _equal_measurements(
     if op1.mv is not None and op2.mv is not None:
         if isinstance(op1.mv, MeasurementValue) and isinstance(op2.mv, MeasurementValue):
-            return qml.equal(op1.mv, op2.mv)
+            mv_equal = qml.equal(op1.mv, op2.mv)
+            if isinstance(mv_equal, str):
+                return mv_equal
+            return mv_equal
```
So we know that `equal` itself always returns a boolean, so we would never hit the `str` branch.

I don't think we need to add too detailed of a message if the measurement values are not the same, but we should probably still return a message like `op1 and op2 are not equal because they have different measurement values` in the case that the measurement values are not equal.

albi3ro commented on pennylane/ops/functions/equal.py:
```diff
@@ -658,41 +658,54 @@ def _equal_measurements(
+        return "MeasurementValue attributes are not iterable or not compatible"
 
-        return False
+    if op1.wires == op2.wires:
+        return True
+    return f"Measurement wires differ: {op1.wires} vs {op2.wires}"
```
```suggestion
    if op1.wires != op2.wires:
        return f"Measurement wires differ: {op1.wires} vs {op2.wires}"
```

If the wires are equal, we need to continue on to the part where we check the eigenvalues.

albi3ro commented on pennylane/ops/functions/equal.py:
```diff
@@ -658,41 +658,53 @@ def _equal_measurements(
+            if qml.math.allclose(op1.eigvals(), op2.eigvals(), rtol=rtol, atol=atol):
+                return True
+            return f"Eigenvalues are not close: op1.eigvals={op1.eigvals()}, op2.eigvals={op2.eigvals()}, with rtol={rtol}, atol={atol}"
+        elif op1.eigvals() is None and op2.eigvals() is None:
```
```suggestion
        if op1.eigvals() is None and op2.eigvals() is None:
```
This should keep codefactor happy.

albi3ro commented on tests/ops/functions/test_equal.py:
Mind fleshing out the tests a it more? The changed code hit by the corresponding `qml.equal` tests, but we aren't actually checking how they integrate with `qml.assert_equal`. 

One good way to do this might be go through `TestMeasurementsEqual`:

https://github.com/PennyLaneAI/pennylane/blob/1433e71d238a8bf8d3edca149a6569f6f1c4a6ef/tests/ops/functions/test_equal.py#L1401

and for every `assert not qml.equal` check, write a :
```
with pytest.raises(AssertionError, match=important_part_of_message):
    qml.assert_equal ...
```
equivalence. That might be a good way to make sure the `assert_equal` tests hit all the relevant branches.

andrijapau commented:
Thanks @tanuj-rai ! @albi3ro has a great suggestion for testing ([here](https://github.com/PennyLaneAI/pennylane/pull/7421#discussion_r2248121674)). Let me / us know if it is unclear and we can help. 🙂 

Apologies for the late reply, this PR went under my radar for quite some time. In the future, could you please re-tag us using the button at the top of the PR? This ensures it enters my list of PRs that require review - ensuring a timely response. 

<img width="392" height="133" alt="image" src="https://github.com/user-attachments/assets/dbf2d9f8-7c82-49a9-b64f-190e9a53e2df" />

Thank you so much for your effort! Let's try and get this across the line 😎 .

tanuj-rai commented:
> Thanks @tanuj-rai ! @albi3ro has a great suggestion for testing ([here](https://github.com/PennyLaneAI/pennylane/pull/7421#discussion_r2248121674)). Let me / us know if it is unclear and we can help. 🙂
> 
> Apologies for the late reply, this PR went under my radar for quite some time. In the future, could you please re-tag us using the button at the top of the PR? This ensures it enters my list of PRs that require review - ensuring a timely response.
> 
> <img alt="image" width="392" height="133" src="https://private-user-images.githubusercontent.com/46359773/473494864-dbf2d9f8-7c82-49a9-b64f-190e9a53e2df.png?jwt=eyJ0eXAiOiJKV1QiLCJhbGciOiJIUzI1NiJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3NTgxOTE3MTgsIm5iZiI6MTc1ODE5MTQxOCwicGF0aCI6Ii80NjM1OTc3My80NzM0OTQ4NjQtZGJmMmQ5ZjgtN2M4Mi00OWE5LWI2NGYtMTkwZTlhNTNlMmRmLnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNTA5MTglMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjUwOTE4VDEwMzAxOFomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWQ2NmYwMjY2ODgwYWVkYTRmZGMzMjNjZWQ3Nzk4NzliN2RjNGM1Njk4NWMzNzdmNDU3OGIwZTcwNTE4MWE2ZTImWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.tcyogjRyd8Np-8EpD49IpMnmYJQZMHhEBNG6JzcutIg">
> Thank you so much for your effort! Let's try and get this across the line 😎 .

Hi @andrijapau, apologies for the really long delay in getting back to this. I have now updated the tests as @albi3ro suggested. Please let me know if anything else needs to be refined. Happy to make further changes. Thank you so much for your patience and guidance!

andrijapau commented on doc/releases/changelog-dev.md:
```diff
@@ -733,9 +733,28 @@
+* Updated `qml.equal` dispatch for `MeasurementProcess` to return informative string messages when observables are not equal.
+ [(#7421)](https://github.com/PennyLaneAI/pennylane/pull/7421)
+
+* Updated requirements and pyproject files to include the other package.  
```
Looks like you have some merge conflicts in the changelog here. Do you mind fixing it?

andrijapau commented on tests/ops/functions/test_equal.py:
```diff
@@ -1680,6 +1722,22 @@ def test_mid_measure(self):
+        with pytest.raises(AssertionError, match="reset"):
+            qml.assert_equal(mp, mp2)
+
+        with pytest.raises(AssertionError, match="id"):
```
Do you mind updating the `match=` to have a more descriptive message rather than a single word?

andrijapau commented:
Hi @tanuj-rai , I hope you're having a good week. 

Since this PR has been inactive for a while, I'll be closing it to keep our list of open PRs manageable. 

Please feel free to reopen it whenever you're ready to continue working on it!
